### PR TITLE
Fix Broken Numeric Shard Generations in RepositoryData

### LINKED ISSUE #57798: Snapshot Repositories Containing a Mix of pre and post v7.6 Snapshots Can Become Corrupted
Repositories that contain both snapshots from before version 7.6 and after 7.6 can become dysfunctional and in some cases corrupted by ES v7.7 clusters as a result of a mistake in how `RepositoryData` is cached. 
The `RepositoryData` is cached including `ShardGenerations` that include numeric generation values that might not be reliable (any failed snapshot finalization that had at least one individual shard snapshot would cause an incorrect shard generation to be tracked).

This leads to two stages of broken behavior:

1. As long as there is still at least one pre-7.6 snapshot in the repository, new `ShardGenerations` will not be physically written to the repository. The issue will show up in errors like while creating new snapshots of affected shards, leading to `PARTIAL` snapshots because the affected shards will never successfully snapshot. 

```
 [2020-06-04T00:00:00.206Z][WARN][org.elasticsearch.snapshots.SnapshotShardsService] [instance-0000000012] [[xxx][0]][xxx:xxx/Lnmw3145RGubRA7oiWqUsg] failed to snapshot shard
java.nio.file.NoSuchFileException: Blob [snapshots/585b4ab8ad5e44d8a8144df80846222b/indices/IzG2oACDQbiD_1479qE4IA/0/index-7866] does not exist
```

Also, snapshot deletes will log the same error, but will work otherwise. This leads to the second stage of the issue described below.
At this stage of the problem, the repository can be fixed and further corruption prevented by setting the setting the repository setting `cache_repository_data` to `false`.

2. Once all the pre-7.6 snapshots have been deleted from a repository the broken `RepositoryGenerations` that were incorrectly cached, will be written to the repository physically.
Once this has happened the repository is physically corrupted and the only way to fix it at this point is to delete all snapshots referencing the broken shards.

We will do two steps of fixing things here:

- [x] Fix the `RepositoryData` caching #57785 
- [ ] Add logic to automatically fix affected `RepositoryData` so that upgrading restores the affected repositories to full functionality

cc @ywelsch , @paulcoghlan

### PR DESCRIPTION
**Pull request by original-brownbear**
Fix broken numeric shard generations when reading them from the wire
or physically from the physical repository.
This seems to me to be the cheapest way to clean up broken shard generations
in a BwC and safe-to-backport manner for now. We can potentially
further optimize this by also not doing the checks on the generations
based on the versions we see in the `RepositoryData` but I don't think
it matters much since we will read `RepositoryData` from cache in almost
all cases.

Closes #57798

### CODE REVIEW

**Comment by original-brownbear** [comment]
Hold off one second with reviewing, just thought of a problematic corner case that needs a (simple) fix still

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 1]
```diff
@@ -615,7 +619,9 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
                         assert indexId != null;
                         indexSnapshots.put(indexId, Collections.unmodifiableList(snapshotIds));
                         for (int i = 0; i < gens.size(); i++) {
-                            shardGenerations.put(indexId, i, gens.get(i));
+                            final String parsedGen = gens.get(i);
+                            shardGenerations.put(
+                                    indexId, i, fixBrokenShardGens ? ShardGenerations.fixShardGeneration(parsedGen) : parsedGen);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:624]: I think we should avoid putting null values into the map (in particular as `ShardGenerations.Builder.put` does not seem to define `@Nullable String generation`, and the semantics of putting null into map can be problematic, as the Map interface says the behavior is underspecified whether a NPE is thrown). I would prefer a boolean method here with conditional logic

File: test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java [hunk 2]
```diff
@@ -129,7 +129,7 @@ public static void assertConsistency(BlobStoreRepository repository, Executor ex
                 try (InputStream blob = blobContainer.readBlob("index-" + latestGen);
                      XContentParser parser = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
                          LoggingDeprecationHandler.INSTANCE, blob)) {
-                    repositoryData = RepositoryData.snapshotsFromXContent(parser, latestGen);
+                    repositoryData = RepositoryData.snapshotsFromXContent(parser, latestGen, true);
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java:132]: why `true` here?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks Armin, I've left one stylistic comment, and one question

File: test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java (hunk 2 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java]: Right my bad no reason to be lenient here ... moved to `false`.

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java]: ++ done

**Comment by original-brownbear** [comment]
Thanks @ywelsch => I pushed https://github.com/elastic/elasticsearch/pull/57813/commits/655c7f3d9f3753a68f2c98718d51ea812fd0f0fd

**Review by ywelsch — APPROVED** [review]
Should we also get this into `7.7.2` (if we ever going to release that one)?

**Comment by original-brownbear** [comment]
@ywelsch I was going to ask the same about the other PR as well ... I don't see why not? => I'll backport this and the other one to 7.7.2?

**Comment by ywelsch** [comment]
👍

**Comment by kostasb** [comment]
Hey @original-brownbear it looks like we missed mentioning this PR in the release notes.
It was backported to 7.7 but we never released 7.7.2, and it's also merged into 7.8 onwards. But I can't find it in the release notes of any version - should it be added to 7.8.0 release notes?

**Comment by original-brownbear** [comment]
@kostasb 

>  should it be added to 7.8.0 release notes?

Yea I guess technically that would be correct. Not sure how relevant it is at this point, but technically it's the first version released with this fix.

**Pull request merged**